torch/_dynamo/guards.py

```diff
@@ -468,11 +469,54 @@ def TENSOR_MATCH(self, guard: Guard):
                     real_value = self.get(tensor_name + "." + term)
                     code.append(f"{tensor_name}.{term} == {real_value}")
-                self._produce_guard_code(guard, code)
             else:
                 self.tensor_check_names.append(tensor_name)
                 self.tensor_check_examples.append(value)
 
+            # A frame is valid for reuse with dynamic dimensions if the new dynamic dimensions are a
+            # strict subset of the old.
+            #
+            # The logic here is as follows:
+            #
+            # Every mark_dynamic directive is a user-knows-best command, which can incur a raise at tracing
+            # time if we find guards that run counter to the user directive.
+            #
+            # If the frame is compiled with any marked dynamic indices, let's call that set of indices X.
+            # When we evaluated inputs against the guards, given the same tensor with potentially new dynamic indices, let's call that set Y.
+            # Y must be a strict subset of X.
+            #
+            # This is the case because any newly introduced mark_dynamic directives have a chance of
+            # incurring raises, failing compilation. Any existing mark_dynamic indices that we lost are safe to lose
+            # as all it means is that we have gotten rid of a user directive which could incur a raise at compile time.
+            # In the case of when there is no Y, that is, there are no dynamic indices marked at all, the frame is safe to reuse
+            # as an empty set is a safe degeneration - that is, a strictly static tensor is always valid for a frame compiled with that same
+            # tensor + more onerous user directives.
+            static, message = tensor_shape_should_be_static(
+                value, guard.source, is_tensor=True
+            )
+            if not static:
+                if hasattr(value, "_dynamo_dynamic_indices"):
+                    # Note - Arguably - if a user marked_dynamic on a dim, but we have *no other* guards on it,
+                    # protecting the directive is inconsequential - and therefore accumulating this guard is spurious.
+                    # TODO(voz): Consider checking for if there are any references to this tensor's dim in guards
+                    # and only guard here if there are.
+                    code.append(
+                        f"({tensor_name}._dynamo_dynamic_indices.issubset({value._dynamo_dynamic_indices})) if hasattr({tensor_name}, '_dynamo_dynamic_indices') else True"
+                    )
+                # In the case of us not having any dynamic dimension indices, we compiled the frame with no chance of
+                # raising for this specific tensor - and any inputs with more dynamic user directives specified must be recompiled.
+                else:
+                    code.append(
+                        f"hasattr({tensor_name}, '_dynamo_dynamic_indices') == False"
```

Comment:
Is the reason for the static test to avoid performing this test uselessly on parameters? Now that we have safety, I'm kind of wondering about perf: post this PR, we now have to do a hasattr test on every tensor, even when no dynamic dims are used. Do you think this is going to be OK?

Reply:
Yes, that is part of it - to just avoid useless tests. 

> we now have to do a hasattr test on every tensor, even when no dynamic dims are used, Do you think this is going to be OK?

I am not sure - I may move it to the C++ guards, I have to benchmark it.

Reply:
Benchmark sweep on eager with both dyn and without looks almost identical, sometimes one is faster, sometimes the other. Jitter aside, this looks fine. Added a comment to that affect either way, in case we do identify this as a problem.